### ISSUE TITLE
Occlusion Culling infective?

### ISSUE THREAD

**Issue by lhofhansl** [minetest/minetest#4722]
I was looking into why occlusion culling only culls relatively few blocks away.

```
			v3s16 cpn = block->getPos() * MAP_BLOCKSIZE;
			cpn += v3s16(MAP_BLOCKSIZE / 2, MAP_BLOCKSIZE / 2, MAP_BLOCKSIZE / 2);
			float step = BS * 1;
			float stepfac = 1.1;
			float startoff = BS * 1;
			float endoff = -BS*MAP_BLOCKSIZE * 1.42 * 1.42;
			v3s16 spn = cam_pos_nodes + v3s16(0, 0, 0);
			s16 bs2 = MAP_BLOCKSIZE / 2 + 1;
			u32 needed_count = 1;
			if (occlusion_culling_enabled &&
					isOccluded(this, spn, cpn + v3s16(0, 0, 0),
						step, stepfac, startoff, endoff, needed_count, nodemgr) &&
```

Does somebody know why we need this: `float endoff = -BS*MAP_BLOCKSIZE * 1.42 * 1.42;`?

If my math is correct, we're off by a factor of 2. We're aliasing all blocks to the _center_ of the enclosing mapblock, so the error (i.e. a block in the same mapblock obscuring the block in question) is MAP_BLOCKSIZE / 2 * 1.42 * 1.42 or simply just block MAP_BLOCKSIZE.

So we should be able to set `float endoff = -BS * MAP_BLOCKSIZE` (-160, as opposed to -320).

This increases the number of culled blocks significantly. I did not detect any rendering issues with this.

Or perhaps I am missing something...? Or perhaps even simpler, we can terminate also when we reach the block mapblock.

**Comment by paramat** [comment]
Erm, i get the feeling 'endoff' is set to the maximum diagonal of a mapblock so that for every point checked (mapblock centre and 8 vertices) the search always terminates before reaching the mapblock. Perhaps ideally 'endoff' would be set to a different value for each point checked but perhaps it's more optimal to not do this.
Not sure about any of this though :]

**Comment by paramat** [comment]
My guess is: the search must terminate before the mapblock so that no nodes in the mapblock are checked. If occlusion was decided as positive depending on a node inside the mapblock, then the whole mapblock would be removed from rendering, but that would remove the node that decided occlusion and remove nodes within the mapblock that were searched through before occlusion was decided.

**Comment by paramat** [comment]
However endoff can be reduced to half it's normal value for the centre point of the mapblock.
All other points need the current endoff because they are being searched towards from all angles, so the max diagonal has to be used.

**Comment by lhofhansl** [comment]
I have PR for something like this with #4724 (also includes a reduction of what is sent from the server).

endoff/2 is essentially what I saying (1/2 of the cube's diagonal). Now in my playing I found one glitch, which I believe to be an off-by-one... Was fixed when I did endOff = -BS*(MAX_BLOCKSIZE+1) (or 170). I guess at the very far corner we could _just_ be in the current map block and incorrectly consider a block culled.
